File: ServerCode/src/Utility/Flag/TypeFlag.ts
export enum UserStatus {
   Student, Teacher, Guest
}

export interface EnvironmentType {
    users : Map<string, UserComponentType>
}

export interface UserComponentType {
    socket_id : string
    name : string,
    user_id : string,
    room_id : string,
    type : UserStatus,
    connection : boolean,
    
    //Only teacher might have this value
    mobilephone? : number,
}

export interface RoomComponentType {
    room_id : string,
    host_id : string,

    start_time : number,
    end_time : number,

    //SocketID List
    students : string[]
}

export interface MessageType {
    event_id : string,
    socket_id : string,
    string_message? : string,
    bool_message? : boolean,
    num_message? : number
}

export const SocketIOKey = {
    socket_id : "socket_id",
    roomCapacity : "room_capacity",

    teacherType : "teacher",
    studentType : "student",
    guestType : "guest",
}

export interface TeacherCreateMsgRoomType {
    user_id : string,
    room_id : string,
    root_socket_id : string
}

export interface TeacherCommonType {
    room_id : string
}

export interface LoginReturnType { 
    status : boolean,
    username? : string,
    user_id? : string,
    seat? : string,
    room_id? : string
}

export interface DatabaseResultType {
    status : boolean,
    result? : any
}

export interface UserDataType {
    userType : UserStatus,
    socket_id : string,
    user_id : string,
    user_name : string,
    room_id? : string
}

export interface RoomStudentType {
    socketID : string,
    user_id : string
}

export interface TerminateEventType {
    location_id: string,
    room_id : string
}

export interface ReserveUserType {
    socket_id : string,
    root_id : string,
    startReserveTime : number
}

export interface AccessTokenType {
    socket_id : string
}

export interface ReconnectRequestType
{
    reconnect_sid: string;
    target_sid : string;
}

export const ClassCSVKey = {
    id : "序號",
    year : "學年",
    semester : "學期",
    class_id : "班級代碼",
    class_name : "班級名稱",
    grade : "年級"
}

File: HCVSAR_UnityProject/Assets/Hsinpa/Script/Utility/TypeFlag.cs

public class TypeFlag
{
    public enum UserType {Student, Teacher, Guest };


    public class SocketEvent {
        public const string OnConnect = "OnConnect";
        public const string UpdateUserInfo = "event@update_userInfo";
        public const string UserJoined = "event@user_joined";
        public const string RefreshUserStatus = "event@refresh_user_status";
        public const string UserLeaved = "event@user_leave";
        public const string CreateRoom = "event@create_room";
        public const string StartGame = "event@start_game";
        public const string TerminateGame = "event@force_end_game";
        public const string KickFromGame = "event@kick_from_game";
    }

    public class InGameType {
        [System.Serializable]
        public struct MissionType
        {
            public string mission_id;
            public string mission_name;
        }
    }

    public class SocketDataType
    {
        [System.Serializable]
        public struct UserDataType
        {
            public TypeFlag.UserType userType;
            public string user_id;
            public string user_name;
            public string room_id;
        }

        [System.Serializable]
        public struct LoginDatabaseType
        {
            public string username;
            public string user_id;
            public string room_id;
            public string seat;

            public TypeFlag.UserType userType;
        }

        [System.Serializable]
        public struct LoginDataStruct
        {
            public TypeFlag.UserType type;
            public string account;
            public string password;
        }

        [System.Serializable]
        public struct RegisterDataStruct
        {
            public string account;
            public string name;
            public string class_id;
        }

        [System.Serializable]
        public struct GeneralDatabaseType
        {
            public bool status;
            public string result;
        }

        [System.Serializable]
        public struct ClassroomDatabaseType
        {
            public string class_name;
            public string class_id;
            public int grade;
            public int year;
        }

        [System.Serializable]
        public struct TeacherCreateMsgRoomType
        {
            public string user_id;
            public string room_id;
        }

        [System.Serializable]
        public struct StudentDatabaseType
        {
            public string id;
            public int year;
            public int semester;
            public string student_name;
            public string seat;
            public string class_id;
        }

        [System.Serializable]
        public struct UserComponentType
        {
            public string socket_id;
            public string name;
            public string user_id;
            public string room_id;
            public TypeFlag.UserType type;
            public int mobilephone;
        }

        [System.Serializable]
        public struct UserScoreType
        {
            public int id;
            public string user_id;
            public string mission_id;
            public int score;
        }

        [System.Serializable]
        public struct RoomComponentType
        {
            public string room_id;
            public string host_id;
            public long start_time;
            public long end_time;

            //public string[] students;
        }

        [System.Serializable]
        public struct ClassScoreHolderType {
            public ClassScoreType[] participant_count;
            public ClassScoreType[] average_score;
        }

        [System.Serializable]
        public struct ClassScoreType
        {
            public string mission_id;
            public float main_value;
        }


    }
}
